plugin/pkg/shoot/validator/admission.go

```diff
@@ -1061,15 +1095,41 @@ func validateKubernetesVersionConstraints(constraints []core.ExpirableVersion, s
 	return false, defaultToLatestPatchVersion, validValues, nil
 }
 
-func validateMachineTypes(constraints []core.MachineType, machine, oldMachine core.Machine, regions []core.Region, region string, zones []string) (bool, []string) {
+func validateMachineTypes(constraints []core.MachineType, machine, oldMachine core.Machine, regions []core.Region, region string, zones []string) (bool, bool, bool, bool, []string) {
 	if machine.Type == oldMachine.Type && pointer.StringEqual(machine.Architecture, oldMachine.Architecture) {
-		return true, nil
+		return true, true, true, true, nil
 	}
 
-	validValues := []string{}
+	var (
+		isMachinePresentInCloudprofile    = false
+		machinesWithSupportedArchitecture = sets.New[string]()
+		machinesAvailableInAllZones       = sets.New[string]()
+		usableMachines                    = sets.New[string]()
+	)
 
-	var unavailableInAtLeastOneZone bool
-top:
+	for _, t := range constraints {
+		if pointer.StringEqual(t.Architecture, machine.Architecture) {
+			machinesWithSupportedArchitecture.Insert(t.Name)
+		}
+		if pointer.BoolDeref(t.Usable, false) {
+			usableMachines.Insert(t.Name)
+		}
+		if !isUnavailableInAtleastOneZone(regions, region, zones, t.Name, "") {
+			machinesAvailableInAllZones.Insert(t.Name)
+		}
+		if t.Name == machine.Type {
+			isMachinePresentInCloudprofile = true
+		}
+	}
+
+	return isMachinePresentInCloudprofile,
+		machinesWithSupportedArchitecture.Has(machine.Type),
+		machinesAvailableInAllZones.Has(machine.Type),
+		usableMachines.Has(machine.Type),
+		machinesWithSupportedArchitecture.Intersection(machinesAvailableInAllZones).Intersection(usableMachines).UnsortedList()
+}
+
+func isUnavailableInAtleastOneZone(regions []core.Region, region string, zones []string, machineType, volumeType string) bool {
```

Comment:
Would it not be cleaner to have a type and a function as parameters so that we do not need to check if one of the types are empty?

Something along the lines:

```
func isUnavailableInAtleastOneZone(regions []core.Region, region string, zones []string, type string, types func(zone core.AvailabilityZone) []string)
```

Reply:
Much better, thanks!